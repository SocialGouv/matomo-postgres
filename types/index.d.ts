import './missing'

import { ColumnType } from 'kysely'

export * from './matomo-api'

export type MatomoTableName = 'matomo' | string

export interface MatomoTable {
  idsite: string
  idvisit: string
  actions?: string
  country?: string
  region?: string
  city?: string
  operatingsystemname?: string
  devicemodel?: string
  devicebrand?: string
  visitduration?: string
  dayssincefirstvisit?: string
  visitortype?: string
  sitename?: string
  userid?: string
  resolution?: string
  serverdateprettyfirstaction?: ColumnType<Date>
  action_id: string
  action_type: string
  action_eventcategory: string
  action_eventaction: string
  action_eventname: string
  action_eventvalue: number
  action_timespent: string
  action_timestamp: ColumnType<Date>
  usercustomproperties?: Record<string, any>
  usercustomdimensions?: Record<string, any>
  dimension1?: string
  dimension2?: string
  dimension3?: string
  dimension4?: string
  dimension5?: string
  dimension6?: string
  dimension7?: string
  dimension8?: string
  dimension9?: string
  dimension10?: string
  action_url?: string
  sitesearchkeyword?: string
  action_title?: string
  visitorid?: string
  referrertype?: string
  referrername?: string
}
