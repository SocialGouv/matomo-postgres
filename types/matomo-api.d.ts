export type Visits = Visit[]

export type Visit = {
  idSite: string
  idVisit: string
  visitIp: any
  visitorId: string
  fingerprint: boolean
  actionDetails: ActionDetail[]
  goalConversions: number
  siteCurrency: string
  siteCurrencySymbol: string
  serverDate: string
  visitServerHour: string
  lastActionTimestamp: number
  lastActionDateTime: string
  siteName: string
  serverTimestamp: number
  firstActionTimestamp: number
  serverTimePretty: string
  serverDatePretty: string
  serverDatePrettyFirstAction: string
  serverTimePrettyFirstAction: string
  userId: any
  visitorType: string
  visitorTypeIcon: string
  visitConverted: string
  visitConvertedIcon?: string
  visitCount: string
  visitEcommerceStatus: string
  visitEcommerceStatusIcon?: string
  daysSinceFirstVisit: number
  secondsSinceFirstVisit: string
  daysSinceLastEcommerceOrder: number
  secondsSinceLastEcommerceOrder: string
  visitDuration: string
  visitDurationPretty: string
  searches: string
  actions: string
  interactions: string
  referrerType: string
  referrerTypeName: string
  referrerName: string
  referrerKeyword: string
  referrerKeywordPosition: any
  referrerUrl?: string
  referrerSearchEngineUrl?: string
  referrerSearchEngineIcon?: string
  referrerSocialNetworkUrl?: string
  referrerSocialNetworkIcon?: string
  languageCode: string
  language: string
  deviceType: string
  deviceTypeIcon: string
  deviceBrand: string
  deviceModel: string
  operatingSystem: string
  operatingSystemName: string
  operatingSystemIcon: string
  operatingSystemCode: string
  operatingSystemVersion: string
  browserFamily: string
  browserFamilyDescription: string
  browser: string
  browserName: string
  browserIcon: string
  browserCode: string
  browserVersion: string
  totalEcommerceRevenue: string
  totalEcommerceConversions: string
  totalEcommerceItems: string
  totalAbandonedCartsRevenue: string
  totalAbandonedCarts: string
  totalAbandonedCartsItems: string
  events: string
  continent: string
  continentCode: string
  country: string
  countryCode: string
  countryFlag: string
  region?: string
  regionCode?: string
  city?: string
  location: string
  latitude: string
  longitude: string
  visitLocalTime: string
  visitLocalHour: string
  daysSinceLastVisit: number
  secondsSinceLastVisit: string
  resolution: string
  plugins: string
  pluginsIcons?: PluginsIcon[]
  dimension1?: string
  dimension2: string
  dimension3?: string
  dimension4?: string
  dimension5?: string
  dimension6?: string
  dimension7?: string
  dimension8?: string
  dimension9?: string
  dimension10?: string
  experiments: Experiment[]
  customVariables: any
  formConversions: number
  sessionReplayUrl: any
  campaignId: string
  campaignContent: string
  campaignKeyword: string
  campaignMedium: string
  campaignName: string
  campaignSource: string
  campaignGroup: string
  campaignPlacement: string
}

export interface ActionDetail {
  type: string
  url?: string
  pageTitle?: string
  pageIdAction?: string
  idpageview?: string
  serverTimePretty: string
  pageId?: string
  pageviewPosition?: string
  title: string
  subtitle?: string
  icon: string
  iconSVG?: string
  timestamp: number
  dimension1?: string
  dimension2?: string
  dimension3?: string
  dimension4?: string
  dimension5?: string
  formName?: string
  formId?: string
  formStatus?: string
  converted?: string
  submitted?: number
  timeToFirstSubmission?: string
  timeSpent: any
  timeHesitation?: string
  leftBlank?: number
  fields?: Field[]
  eventCategory?: string
  eventAction?: string
  timeSpentPretty?: string
  eventName?: string
  eventValue?: number
  revenue?: any
  items?: string
  itemDetails?: ItemDetail[]
  customVariables?: CustomVariables
  server_time?: string
  goalName?: string
  goalId?: string
  goalPageId?: string
  orderId?: string
  revenueSubTotal?: number
  revenueTax?: number
  revenueShipping?: number
  revenueDiscount?: number
  siteSearchKeyword: string
}

export interface Field {
  fieldName: string
  timeSpent: string
  timeHesitation: string
  leftBlank: string
  submitted: string
}

export interface ItemDetail {
  itemSKU: string
  itemName: string
  itemCategory: string
  price: any
  quantity: string
  categories: string[]
}

export interface CustomVariables {
  [key: string]: N1 | N2 | N3 | N4 | N5 | N6 | N7 | N8 | N9
}

export interface N1 {
  customVariablePageName1: string
  customVariablePageValue1: string
}

export interface N2 {
  customVariablePageName2: string
  customVariablePageValue2: string
}

export interface N3 {
  customVariablePageName3: string
  customVariablePageValue3: string
}

export interface N4 {
  customVariablePageName4: string
  customVariablePageValue4: string
}
export interface N5 {
  customVariablePageName5: string
  customVariablePageValue5: string
}
export interface N6 {
  customVariablePageName6: string
  customVariablePageValue6: string
}
export interface N7 {
  customVariablePageName7: string
  customVariablePageValue7: string
}
export interface N8 {
  customVariablePageName8: string
  customVariablePageValue8: string
}
export interface N9 {
  customVariablePageName9: string
  customVariablePageValue9: string
}

export interface PluginsIcon {
  pluginIcon: string
  pluginName: string
}

export interface Experiment {
  idexperiment: string
  name: string
  variation: Variation
}

export interface Variation {
  idvariation: any
  name: string
}
