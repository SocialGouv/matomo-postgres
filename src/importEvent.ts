import { InsertResult } from "kysely";
import { ActionDetail, Visit } from "types/matomo-api";

import { DESTINATION_TABLE } from "./config";
import { db } from "./db";

/**
 *
 * @param {Client} client
 * @param {import("types").Event} event
 *
 * @return {Promise<Record<"rows", any[]>>}
 */
export const importEvent = (event: MatomoActionDetail): Promise<InsertResult[]> =>
  // @ts-ignore // TODO
  db
    .insertInto(DESTINATION_TABLE)
    .values([{ ...event }])
    .onConflict((oc) => oc.doNothing())
    .execute();

const matomoProps = [
  "idSite",
  "idVisit",
  "actions",
  "country",
  "region",
  "city",
  "operatingSystemName",
  "deviceModel",
  "deviceBrand",
  "visitDuration",
  "daysSinceFirstVisit",
  "visitorType",
  "visitorId",
  "referrerType",
  "referrerName",
  "siteName",
  "userId",
  "resolution",
] as const;

/** @type Record<string, (a: import("types/matomo-api").ActionDetail) => string | number> */
const actionProps = {
  action_type: (action: ActionDetail) => action.type,
  action_title: (action: ActionDetail) => action.title,
  action_eventcategory: (action: ActionDetail) => action.eventCategory,
  action_eventaction: (action: ActionDetail) => action.eventAction,
  action_eventname: (action: ActionDetail) => action.eventName,
  action_eventvalue: (action: ActionDetail) => action.eventValue,
  action_timespent: (action: ActionDetail) => action.timeSpent,
  action_timestamp: (action: ActionDetail) => new Date(action.timestamp * 1000).toISOString(),
  action_url: (action: ActionDetail) => action.url,
  sitesearchkeyword: (action: ActionDetail) => action.siteSearchKeyword,
};

export const getEventsFromMatomoVisit = (matomoVisit: Visit): MatomoActionDetail[] => {
  return matomoVisit.actionDetails.map((actionDetail, actionIndex) => {
    const usercustomproperties: Record<string, any> = {};
    for (let k = 1; k < 10; k++) {
      const property = actionDetail.customVariables && actionDetail.customVariables[k];
      if (!property) continue; // max 10 custom variables
      //@ts-ignore
      usercustomproperties[property[`customVariableName${k}`]] = property[`customVariableValue${k}`];
    }

    /** @type {Record<string, string>} */
    const usercustomdimensions = {};
    for (let k = 1; k < 11; k++) {
      const dimension = `dimension${k}`;
      //@ts-ignore
      const value = actionDetail[dimension] || matomoVisit[dimension];
      if (!value) continue; // max 10 custom variables
      //@ts-ignore
      usercustomdimensions[dimension] = value;
    }

    const event = {
      // default matomo visit properties
      ...matomoProps.reduce((a, prop) => ({ ...a, [prop.toLowerCase()]: matomoVisit[prop] }), {}),
      serverdateprettyfirstaction: new Date((matomoVisit.firstActionTimestamp || 0) * 1000).toISOString(),
      // action specific properties
      ...Object.keys(actionProps).reduce(
        (a, prop) => ({
          ...a,
          [prop.toLowerCase()]: actionProps[prop as ActionPropsKeys](actionDetail),
        }),
        {
          action_id: `${matomoVisit.idVisit}_${actionIndex}`,
        }
      ),
      // custom variables
      usercustomproperties,
      // custom dimensions
      // We keep both for backwards compatibility.
      // Current implementation is flat with one column for each dimension.
      usercustomdimensions,
      ...usercustomdimensions,
    };
    return event;
  });
};

type ActionPropsKeys = keyof typeof actionProps;

type AllMatomoActionDetailKeys =
  | Lowercase<(typeof matomoProps)[number]>
  | Lowercase<ActionPropsKeys>
  | "dimension1"
  | "dimension2"
  | "dimension3"
  | "dimension4"
  | "dimension5"
  | "dimension6"
  | "dimension7"
  | "dimension8"
  | "dimension9"
  | "dimension10"
  | "action_id"
  | "usercustomproperties"
  | "usercustomdimensions"
  | "serverdateprettyfirstaction";

type MatomoActionDetail = Partial<
  Record<AllMatomoActionDetailKeys, string> & { usercustomproperties: any } & {
    usercustomdimensions: any;
  }
>;
