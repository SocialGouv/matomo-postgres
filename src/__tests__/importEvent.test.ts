import { Visit } from 'types/matomo-api'

import { getEventsFromMatomoVisit } from '../importEvent'
import matomoVisit from './visit.json'

process.env.MATOMO_SITE = '42'
process.env.PROJECT_NAME = 'some-project'
process.env.RESULTPERPAGE = '10'

test('getEventsFromMatomoVisit: should merge action events', () => {
  const visits = getEventsFromMatomoVisit(matomoVisit as unknown as Visit)
  expect(visits).toMatchSnapshot()
})
