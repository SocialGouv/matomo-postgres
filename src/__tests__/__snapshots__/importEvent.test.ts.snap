// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEventsFromMatomoVisit: should merge action events 1`] = `
[
  {
    "action_eventaction": "Cart change",
    "action_eventcategory": "Ecommerce",
    "action_eventname": "added - Basic Wetsuit",
    "action_eventvalue": 1,
    "action_id": "124_0",
    "action_timespent": 48,
    "action_timestamp": "2021-08-20T21:35:18.000Z",
    "action_title": "Evènement",
    "action_type": "event",
    "action_url": "https://dive-shop.net/products/basic-wetsuit/",
    "actions": "2",
    "city": "Buenos Aires",
    "country": "Argentine",
    "dayssincefirstvisit": "23",
    "devicebrand": "Inconnu",
    "devicemodel": "Générique Bureau",
    "dimension1": "guest",
    "dimension10": "36",
    "dimension2": "julien",
    "dimension3": "page",
    "dimension4": "indonesia",
    "dimension5": "diving",
    "dimension6": "shop",
    "dimension7": "v1.2.3",
    "dimension8": "fr",
    "dimension9": "light",
    "idsite": "42",
    "idvisit": "124",
    "operatingsystemname": "Mac",
    "referrername": "referrerName",
    "referrertype": "referrerType",
    "region": "Buenos Aires",
    "resolution": "1920x1080",
    "serverdateprettyfirstaction": "2021-08-20T21:55:12.000Z",
    "sitename": "tests",
    "sitesearchkeyword": undefined,
    "usercustomdimensions": {
      "dimension1": "guest",
      "dimension10": "36",
      "dimension2": "julien",
      "dimension3": "page",
      "dimension4": "indonesia",
      "dimension5": "diving",
      "dimension6": "shop",
      "dimension7": "v1.2.3",
      "dimension8": "fr",
      "dimension9": "light",
    },
    "usercustomproperties": {
      "page-author": "Julien",
      "post-age": "-430 days",
    },
    "userid": "24",
    "visitduration": "300",
    "visitorid": "visitorId",
    "visitortype": "returningCustomer",
  },
  {
    "action_eventaction": undefined,
    "action_eventcategory": undefined,
    "action_eventname": undefined,
    "action_eventvalue": undefined,
    "action_id": "124_1",
    "action_timespent": "2",
    "action_timestamp": "2021-08-20T21:30:25.000Z",
    "action_title": "Divezone Brand Diving Boots - Divezone Store",
    "action_type": "action",
    "action_url": "https://dive-shop.net/products/diving-boots/",
    "actions": "2",
    "city": "Buenos Aires",
    "country": "Argentine",
    "dayssincefirstvisit": "23",
    "devicebrand": "Inconnu",
    "devicemodel": "Générique Bureau",
    "dimension1": "guest",
    "dimension10": "36",
    "dimension2": "julien",
    "dimension3": "page",
    "dimension4": "indonesia",
    "dimension5": "diving",
    "dimension6": "shop",
    "dimension7": "v1.2.3",
    "dimension8": "fr",
    "dimension9": "light",
    "idsite": "42",
    "idvisit": "124",
    "operatingsystemname": "Mac",
    "referrername": "referrerName",
    "referrertype": "referrerType",
    "region": "Buenos Aires",
    "resolution": "1920x1080",
    "serverdateprettyfirstaction": "2021-08-20T21:55:12.000Z",
    "sitename": "tests",
    "sitesearchkeyword": undefined,
    "usercustomdimensions": {
      "dimension1": "guest",
      "dimension10": "36",
      "dimension2": "julien",
      "dimension3": "page",
      "dimension4": "indonesia",
      "dimension5": "diving",
      "dimension6": "shop",
      "dimension7": "v1.2.3",
      "dimension8": "fr",
      "dimension9": "light",
    },
    "usercustomproperties": {},
    "userid": "24",
    "visitduration": "300",
    "visitorid": "visitorId",
    "visitortype": "returningCustomer",
  },
  {
    "action_eventaction": undefined,
    "action_eventcategory": undefined,
    "action_eventname": undefined,
    "action_eventvalue": undefined,
    "action_id": "124_2",
    "action_timespent": "2",
    "action_timestamp": "2021-08-20T21:30:22.000Z",
    "action_title": "Divezone Brand Diving Boots - Divezone Store",
    "action_type": "search",
    "action_url": "https://dive-shop.net/products/diving-boots/",
    "actions": "2",
    "city": "Buenos Aires",
    "country": "Argentine",
    "dayssincefirstvisit": "23",
    "devicebrand": "Inconnu",
    "devicemodel": "Générique Bureau",
    "dimension1": "guest",
    "dimension10": "36",
    "dimension2": "julien",
    "dimension3": "page",
    "dimension4": "indonesia",
    "dimension5": "diving",
    "dimension6": "shop",
    "dimension7": "v1.2.3",
    "dimension8": "fr",
    "dimension9": "light",
    "idsite": "42",
    "idvisit": "124",
    "operatingsystemname": "Mac",
    "referrername": "referrerName",
    "referrertype": "referrerType",
    "region": "Buenos Aires",
    "resolution": "1920x1080",
    "serverdateprettyfirstaction": "2021-08-20T21:55:12.000Z",
    "sitename": "tests",
    "sitesearchkeyword": "scuba",
    "usercustomdimensions": {
      "dimension1": "guest",
      "dimension10": "36",
      "dimension2": "julien",
      "dimension3": "page",
      "dimension4": "indonesia",
      "dimension5": "diving",
      "dimension6": "shop",
      "dimension7": "v1.2.3",
      "dimension8": "fr",
      "dimension9": "light",
    },
    "usercustomproperties": {},
    "userid": "24",
    "visitduration": "300",
    "visitorid": "visitorId",
    "visitortype": "returningCustomer",
  },
]
`;
